pallets/eth-bridge/src/lib.rs

```diff
@@ -794,48 +806,62 @@ pub mod pallet {
     fn process_ethereum_events_partition<T: Config>(
         active_range: &ActiveEthRange,
         partition: &EthereumEventsPartition,
-    ) {
+    ) -> Result<(), DispatchError> {
         // Remove entry from storage. Ignore votes.
         let _ = EthereumEvents::<T>::take(partition);
         for discovered_event in partition.events().iter() {
             match ValidEvents::try_from(&discovered_event.event.event_id.signature) {
                 Some(valid_event) =>
                     if active_range.event_types_filter.contains(&valid_event) {
-                        process_ethereum_event::<T>(&discovered_event.event);
+                        process_ethereum_event::<T>(&discovered_event.event)?;
                     } else {
                         log::warn!("Ethereum event signature ({:?}) included in approved range ({:?}), but not part of the expected ones {:?}", &discovered_event.event.event_id.signature, active_range.range, active_range.event_types_filter);
                     },
-                None => log::warn!(
-                    "Unknown Ethereum event signature in range {:?}",
-                    &discovered_event.event.event_id.signature
-                ),
-            };
+                None => {
+                    log::warn!(
+                        "Unknown Ethereum event signature in range {:?}",
+                        &discovered_event.event.event_id.signature
+                    );
+                },
+            }
         }
 
         // Cleanup
         for (partition, votes) in EthereumEvents::<T>::drain() {
             // TODO raise offences
             log::info!("Collators with invalid votes on ethereum events (range: {:?}, partition: {}): {:?}", partition.range(), partition.partition(), votes);
         }
+
+        Ok(())
     }
 
-    fn process_ethereum_event<T: Config>(event: &EthEvent) {
-        // TODO before processing ensure that the event has not already been processed
-        match T::BridgeInterfaceNotification::on_event_processed(&event) {
+    fn process_ethereum_event<T: Config>(event: &EthEvent) -> Result<(), DispatchError> {
+        ensure!(
```

Comment:
we shouldn't error here. If the event has been processed, we avoid double processing, signal/log/emit the appropriate info and return.

Reply:
If we call this function with an event that has already been processed, it's because the upstream functions did not validate properly.

Reply:
if the authors/collators have reached consensus on the tier1 state then we shouldn't block it.
There are valid cases that this can occur:
1. events added and being processed while upgrading.
2. events added to tier2 due starting from an earlier last block
We should definitely mark the case, but not block the event processing mechanism.
The error will cause the last vote to be stuck and recasted, only to be rejected again and again in a loop.

Reply:
In both of these scenarios, and possibly others, I would expect the collators to check first if the event is valid before submitting it (ex: already processed or `0x000`..). 
But in any case, this function is too low level to be required to be infallible. I think we can still throw an error here and catch it on `process_ethereum_events_partition` and log it there and move on to the next event? 
It feels like making an distinction between a bad event (one that is already processed, or is 0 
 - something we should add in the validation i guess :smile: ) and a valid event that fails to be processed makes sense which is what this low level function is doing.

Reply:
I am happy with either solution, although perhaps handling the error within the `process_ethereum_events_partition` function is the best solution.

Reply:
I have pushed a [commit](https://github.com/AventusProtocolFoundation/avn-parachain/pull/390/commits/4b74809faf44e3f2181cea8496b8180eab997bb4) the changes @nahuseyoum suggested @thadouk . This can be reverted if necessary although I believe it should be ok